File: appengine/monorail/static_src/shared/experiments.test.js
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import {assert} from 'chai';
import {isExperimentEnabled, SLO_EXPERIMENT} from './experiments.js';


describe('isExperimentEnabled', () => {
  it('throws error for unknown experiment', () => {
    assert.throws(() =>
      isExperimentEnabled('unknown-exp', {displayName: 'jessan@google.com'}));
  });

  it('returns false if user not in experiment', () => {
    const ineligibleUser = {displayName: 'example@example.com'};
    assert.isFalse(isExperimentEnabled(SLO_EXPERIMENT, ineligibleUser, {}));
  });

  it('returns false if no user provided', () => {
    assert.isFalse(isExperimentEnabled(SLO_EXPERIMENT, undefined, {}));
  });

  it('returns true if user in experiment', () => {
    const eligibleUser = {displayName: 'jessan@google.com'};
    assert.isTrue(isExperimentEnabled(SLO_EXPERIMENT, eligibleUser, {}));
  });

  it('is false if user in experiment has disabled it with URL', () => {
    const eligibleUser = {displayName: 'jessan@google.com'};
    assert.isFalse(isExperimentEnabled(
        SLO_EXPERIMENT, eligibleUser, {'e': '-slo'}));
  });

  it('ignores enabling experiments with URL', () => {
    const ineligibleUser = {displayName: 'example@example.com'};
    assert.isFalse(isExperimentEnabled(
        SLO_EXPERIMENT, ineligibleUser, {'e': 'slo'}));
  });

  it('ignores ineligible users disabling experiment with URL', () => {
    const ineligibleUser = {displayName: 'example@example.com'};
    assert.isFalse(isExperimentEnabled(
        SLO_EXPERIMENT, ineligibleUser, {'e': '-slo'}));
  });

  it('ignores invalid experiments in URL', () => {
    const eligibleUser = {displayName: 'jessan@google.com'};
    // Leading comma, unknown experiment str, empty experiment str in
    // middle, disable_str with no experiment, trailing comma
    assert.isFalse(isExperimentEnabled(
        SLO_EXPERIMENT, eligibleUser, {'e': ',unknown,-slo,,-,'}));
  });

  it('respects last instance when experiment repeated in URL', () => {
    const eligibleUser = {displayName: 'jessan@google.com'};
    assert.isFalse(isExperimentEnabled(
        SLO_EXPERIMENT, eligibleUser, {'e': 'slo,-slo'}));
    assert.isTrue(isExperimentEnabled(
        SLO_EXPERIMENT, eligibleUser, {'e': '-slo,slo'}));
  });
});

File: appengine/cr-buildbucket/test/experiments_test.py
# Copyright 2021 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

from parameterized import parameterized

from testing_utils import testing

import experiments


class CheckInvalidExperimentNames(testing.AppengineTestCase):

  @parameterized.expand([
      ('cool.experiment'),
      ('single'),
      ('luci.use_realms'),
  ])
  def test_valid(self, exp_name):
    self.assertIsNone(experiments.check_invalid_name(exp_name))

  @parameterized.expand([
      ('bad!name'),
      ('not experiment'),
  ])
  def test_bad_name(self, exp_name):
    self.assertTrue(
        experiments.check_invalid_name(exp_name).startswith('does not match')
    )

  @parameterized.expand([
      ('luci.use_ralms'),
  ])
  def test_reserved_name(self, exp_name):
    self.assertTrue(
        experiments.check_invalid_name(exp_name)
        .startswith('unknown experiment has reserved prefix')
    )
